core/src/main/java/com/linecorp/armeria/common/RequestContextAccessor.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.internal.common.RequestContextUtil;
+
+import io.micrometer.context.ContextRegistry;
+import io.micrometer.context.ContextSnapshot;
+import io.micrometer.context.ContextSnapshot.Scope;
+import io.micrometer.context.ThreadLocalAccessor;
+
+/**
+ * This class works with the
+ * <a href="https://docs.micrometer.io/context-propagation/reference/index.html">
+ * Context-propagation</a> library and keep the {@link RequestContext} during
+ * <a href="https://github.com/reactor/reactor-core">Reactor</a> operations.
+ * Get the {@link RequestContextAccessor} to register it to the {@link ContextRegistry}.
+ * Then, {@link ContextRegistry} will use {@link RequestContextAccessor} to
+ * propagate context during the
+ * <a href="https://github.com/reactor/reactor-core">Reactor</a> operations
+ * so that you can get the context using {@link RequestContext#current()}.
+ * However, please note that you should include Mono#contextWrite(ContextView) or
+ * Flux#contextWrite(ContextView) to end of the Reactor codes.
+ * If not, {@link RequestContext} will not be keep during Reactor Operation.
+ */
+public final class RequestContextAccessor implements ThreadLocalAccessor<RequestContext> {
+
+    private static final String KEY = RequestContextAccessor.class.getName();
+
+    /**
+     * The value which obtained through {@link RequestContextAccessor},
+     * will be stored in the Context under this {@code KEY}.
+     * This method will be called by {@link ContextSnapshot} internally.
+     */
+    @Override
+    public Object key() {
+        return KEY;
+    }
+
+    /**
+     * The value which obtained through {@link RequestContextAccessor},
+     * will be stored in the Context under this {@code KEY}.
+     * User can use this method to register {@link RequestContext} to
+     * Reactor Context.
+     */
+
+    public static String accessorKey() {
```

Comment:
@minwoox nim, No it would be better to remain it!

We should register `Key` - `Value` to `Reactor Context`. 
Because `DefaultContextSnapshot` of `Context-Propagation` get value from `Reactor Context` by using `ThreadLocalAccessor`'s Key.
```java
// https://github.com/micrometer-metrics/context-propagation/blob/b42c0049e46121e4e3419762895a26098eae4321/context-propagation/src/main/java/io/micrometer/context/DefaultContextSnapshot.java#L81-L91
            Object key = accessor.key();
            if (keyPredicate.test(key)) {
                if (this.containsKey(key)) {
                    Object value = get(key);
                    assert value != null : "snapshot contains disallowed null mapping for key: " + key;
                    previousValues = setThreadLocal(key, value, accessor, previousValues);
                }
                else if (clearMissing) {
                    previousValues = clearThreadLocal(key, accessor, previousValues);
                }
            }
```

That means that we should know `RequestContextAccssor`'s Key.
If there is no method like `public static String accessorKey() {...}`, there are two options to achieve it.

1. `new RequestContextAccesore().key()`. 
2. `private static final String KEY` -> `public static final String KEY`

IMHO, two options are not suitable. 

```java
        return mono.doFirst(() -> assertThat(ctxExists(ctx)).isTrue())
                   .doOnSubscribe(s -> assertThat(ctxExists(ctx)).isTrue())
                   .doOnRequest(l -> assertThat(ctxExists(ctx)).isTrue())
                   .doOnNext(foo -> assertThat(ctxExists(ctx)).isTrue())
                   .doOnSuccess(t -> assertThat(ctxExists(ctx)).isTrue())
                   .doOnEach(s -> assertThat(ctxExists(ctx)).isTrue())
                   .doOnError(t -> assertThat(ctxExists(ctx)).isTrue())
                   .doAfterTerminate(() -> assertThat(ctxExists(ctx)).isTrue())
                   // Please note this line. RequestContextAccessor.accesorKey().
                   .contextWrite(Context.of(RequestContextAccessor.accessorKey(), ctx));
```


What do you think?

Reply:
> Because DefaultContextSnapshot of Context-Propagation get value from Reactor Context by using ThreadLocalAccessor's Key.

If so, can't we just use `RequestContextAccessor.key()` instead of `RequestContextAccessor.accessorKey()`? 😉 

Now I understand. So, users should manually use `contextWrite` instead of automatically retrieving the `RequestContext` from the thread-local and propagating it. I'm not entirely convinced that this is the best approach for utilizing this Micrometer propagation. Let me spend some more time considering this.

Reply:
> Now I understand. So, users should manually use contextWrite instead of automatically retrieving the RequestContext from the thread-local and propagating it. I'm not entirely convinced that this is the best approach for utilizing this Micrometer propagation. Let me spend some more time considering this.

Yes, as you said, users should manually use `contextWrite`. IMHO, there are dark and light.  😅 

Advantages
1. `microMeter/context-propagation` may support other libraries in the future. In that case, `Armeria` just needs to use it as it is. Additionally, `Reactor` and `MicroMeter: context-propagation` will continue to support tight integration in the future as well. it means `Armeria` developer is not need to worry about it.
2. May improve performance. When used with `micro-meter:context-propagation`, `Reactor` creates minimal wrapper classes for context propagation. This may also provide a performance advantage.
3. Reactor and ContextPropagation are closely collaborating, and users of other frameworks are also likely to become familiar with it.

Disadvantages
1. People using `RequestContextHook` in `Armeria` will need to modify their code when using `micro-meter:Context-Propagation`. If it's a large-scale server, it may be difficult to add contextWrite() at the end of every API.

Reply:
Thanks for the explanation. Now, I understand how micrometer context-propagation works. Let's keep it as is for now.

Reply:
Thanks for your comment 🙇‍♂️ 
Okay, I keep it as is!